### PR TITLE
`use_query` prefetching for ManyToMany and ManyToOne fields

### PR DESCRIPTION
**Pull request by Archmonger**
## Description

Attempt at auto fetching ManyToMany fields

## Checklist:

Please update this checklist as you complete each item:

-   [x] Tests have been included for all bug fixes or added functionality.
-   [x] The `changelog.rst` has been updated with any significant changes, if necessary.
-   [x] GitHub Issues which may be closed by this PR have been linked.

### CODE REVIEW

**Comment by Archmonger** [comment]
@rmorshea This may not be possible. `get_queryset` does not solve the problem, and when calling `all()` on the results they don't get cached.

I think the solution may be documenting that `ManyToMany` field fetches require a separate `use_query` hook call.

**Comment by rmorshea** [comment]
On a side note, I'm observing the following traceback. It seems like `render_view` isn't actually returning an `HttpResponse` object, but rather an unawaited coroutine.

```python
Traceback (most recent call last):
  File "/home/ryan/Repositories/idom-team/django-idom/venv/lib/python3.10/site-packages/asgiref/sync.py", line 472, in thread_handler
    raise exc_info[1]
  File "/home/ryan/Repositories/idom-team/django-idom/venv/lib/python3.10/site-packages/django/core/handlers/exception.py", line 42, in inner
    response = await get_response(request)
  File "/home/ryan/Repositories/idom-team/django-idom/venv/lib/python3.10/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
  File "/home/ryan/Repositories/idom-team/django-idom/src/django_idom/http/views.py", line 52, in view_to_component_iframe
    response["X-Frame-Options"] = "SAMEORIGIN"
TypeError: 'coroutine' object does not support item assignment
```

**Comment by Archmonger** [comment]
Weird, I haven't seen that in my testing environment. Not happening on GH actions either.

**Comment by Archmonger** [comment]
I needed to add a kwarg to allow users to turn off `many_to_many` or `many_to_one` prefetch behavior. It may be memory intensive if using deep nested database structures. However, this has broken the `*args: _Params.args, **kwargs: _Params.kwargs,` type hints.

**Comment by rmorshea** [comment]
Unfortunately, `ParamSpec` does not support this. It was explicitly [left out](https://peps.python.org/pep-0612/#concatenating-keyword-parameters) of the original PEP because it was deemed to be of sufficient complexity to be considered out of scope.

Here's one way we could work around this...

Instead of implementing this caching logic in `use_query` itself, this could be done in a wrapper functions:

```python
use_query(prefetch(my_query, fields=True, related=True), *args, **kwargs)
```

Where `prefetch` is of the form:

```
P = ParamSpec("P")
R = TypeVar("R")

def prefetch(query: Callable[P, R], fields: bool: False, related: bool = False) -> Callable[P, R]: ...
```

I actually quite like this since it doesn't impinge on the `*args, **kwargs` of the `query` function. It could also be used to address https://github.com/idom-team/django-idom/issues/104 since you only need to use prefetch if you're returning models or querysets.

**Comment by rmorshea** [comment]
I just realized that `prefetch` could also be used as a decorator:

```python
@prefetch
def my_query(): ...

@prefetch(related=True)
def my_query(): ...
```

**Comment by Archmonger** [comment]
Some form of prefetching would still need to occur within `use_query` to ensure all the basic fields are accessible.

I personally think it feels the least awkward for them to be parameters within `use_query`, truly unfortunate on that `ParamSpec` limitation... I'll try wrappers out.

**Comment by Archmonger** [comment]
Having it be a decorator is also kind of unfortunate, since I was hoping the default behavior would be to prefetch.

Now that I'm thinking about it, it's not unrealistic to ask Django users to add a `prefetch` parameter to their `models.py`. Would fit Django's "fat models" architecture.

**Comment by rmorshea** [comment]
The type hints for `prefetch` would be:

```python
from typing import ParamSpec, TypeVar, overload

_Params = ParamSpec("_Params")
_Result = TypeVar("_Result")

@overload
def prefetch(
    query: None = ..., *, fields: bool = ..., related: bool = ...
) -> Callable[[Callable[_Params, _Result]], Callable[_Params, _Result]]:
    ...

@overload
def prefetch(
    query: Callable[_Params, _Result], *, fields: bool = ..., related: bool = ...
) -> Callable[_Params, _Result]:
    ...

def prefetch(
    query: Callable[_Params, _Result] | None = None, *, fields: bool = True, related: bool = False
) -> Callable[_Params, _Result] | Callable[[Callable[_Params, _Result]], Callable[_Params, _Result]]:
    ...
```

This allows prefetch to be used as a decorator and with an inline call.

**Comment by Archmonger** [comment]
I attempted to generate an option that is extensible, with the knowledge that `use_query` may some day be merged into core.

**Comment by rmorshea** [comment]
I think I'd like to have some more discussion before implementing anything. I've gathered the relevant issue for discussion here: https://github.com/idom-team/django-idom/issues/113

**Comment by rmorshea** [comment]
I'll have time to take a look at this tomorrow.

**Review by rmorshea — CHANGES_REQUESTED** [review]

File: src/django_idom/types.py
Comment (Archmonger) [reply src/django_idom/types.py:69]: Yes, for example sometimes it's more convenient for me to always have the data in json format rather than as a Django `QuerySet`, so I would likely use a generic postprocessor that serializes the data.

IMO it's prettier to have this built-in to the hook.

We'll have a chance to rework this API to be more akin to `React Query` when we migrate this hook into core.

File: src/django_idom/hooks.py
Comment (Archmonger) [reply src/django_idom/hooks.py:89]: From a Python perspective, this would feel wrong. PEP8 positional args should be "absolute" in positioning. So whatever is the first argument should always be the first arg.

```python
# Query is the first arg in a normal situation
use_query(query, arg1, arg2)

# Now query is the second arg, which means query's position is not absolute
use_query(options, query, arg1, arg2)
```

Comment (rmorshea) [reply src/django_idom/hooks.py:89]: One could make a similar argument about the fact that `arg1` must be shifted in order to add options after `query`. It seems to me that, given we're unable to use a kwarg to pass the options, this rule must be broken one way or another. With this in mind, I'd argue that have options go first is preferable because, with and without options, `query` and its options remain visually grouped.

Comment (Archmonger) [reply src/django_idom/hooks.py:89]: Also in the argument of arg1, that's a single violation. But moving both arg1 and query becomes two violations.

Comment (rmorshea) [reply src/django_idom/hooks.py:89]: This seems like a case where there should probably be one way to do things in order to aid in readability.

> But moving both arg1 and query becomes two violations.

I mean, if putting it first causes N violations, then putting it next causes N-1 (since you must shift `arg2` etc.) which I don't think is a big enough difference to counterbalance the value of grouping `query, *args, **kwargs`.

File: src/django_idom/types.py
Comment (rmorshea) [reply src/django_idom/types.py:69]: > IMO it's prettier to have this built-in to the hook.

This may be the case if `postprocessor` defaults to prefetching, but it has the con that the default options no longer make sense if you say, set `postprocessor=to_json`. Also, if I mistakenly configure `postprocessor_options` while using the `to_json` post processor:

```python
QueryOptions(postprocessor_options={"many_to_many": True}, postprocessor=to_json)
```

 I won't be notified that my `postprocessor_options` will have no effect because `to_json` simply won't interpret them.

File: src/django_idom/hooks.py
Comment (Archmonger) [reply src/django_idom/hooks.py:89]: I tried playing around with it, IMO the API feels very wrong to change to shift the query position around.

Since it's type hinted properly, I don't think there's any ambiguity about the second arg situation.

File: src/django_idom/types.py
Comment (rmorshea) [inline src/django_idom/types.py:57]: If this is a `TypedDict` then users don't need to import this class and can instead declare options using dict literals.

Comment (rmorshea) [reply src/django_idom/types.py:57]: I suppose that having options are the first pos arg doesn't suffer from this issue since it is either a dict or callable.

Comment (Archmonger) [reply src/django_idom/types.py:57]: I think that using an object here still simplifies some things later down the line. For example, if we ever change parameter names, it's good to generate exceptions.

**Review by rmorshea — APPROVED** [review]
We can merge this as-is. However, when it comes time to get this into core I'd like to completely rethink the API. In doing so, I think the user experience can be greatly improved.

File: src/django_idom/types.py
Comment (Archmonger) [reply src/django_idom/types.py]: Maybe the solution here is to convert `postprocessor_options` into kwargs for `postprocessor`.

That way, an exception will be generated for incorrect options.

File: src/django_idom/hooks.py
Comment (Archmonger) [reply src/django_idom/hooks.py:83]: I've moved the `QueryOptions` to the first arg

File: src/django_idom/types.py
Comment (Archmonger) [reply src/django_idom/types.py:72]: Closing this, as I believe strictly typed objects have significant benefits over loosely typed dicts.

**Comment by rmorshea** [comment]
Having that be the default seems alright.

**Comment by Archmonger** [comment]
Do you want to review the docs or should I merge?

**Review by rmorshea — APPROVED** [review]

**Comment by rmorshea** [comment]
Besides that one little nit it looks good.

File: docs/src/features/hooks.md
Comment (Archmonger) [reply docs/src/features/hooks.md:124]: While I was trying to fix this, I discovered mypy does not like our type hint for postprocessor 


`Incompatible types in assignment (expression has type "Callable[[Union[_QuerySet[Any, Any], Model], bool, bool], None]", variable has type "Optional[Postprocessor]")`
```python
class Postprocessor(Protocol):
    def __call__(self, data: Any, **kwargs: Any) -> None:
        ...

@dataclass
class QueryOptions:
    """Configuration options that can be provided to `use_query`."""

    postprocessor: Postprocessor | None = django_query_postprocessor
```

Comment (rmorshea) [reply docs/src/features/hooks.md:124]: `**kwargs: Any` in the protocol says that post processor functions must accept arbitrary kwargs. That is, they must be a callable with a `**` param in them. My recommendation would be to just type the post processor as `Callable[..., None]`. It would be cool if you could do:

```python
P = ParamSpec("P")

class QueryOptions:
    postprocessor: Callable[Concatenate[Any, P], None]
    postprocessor_args: P.args
    postprocessor_kwargs: P.kwargs
```

But I don't think this is possible.

Comment (rmorshea) [reply docs/src/features/hooks.md:124]: I think you could do:
```python
class Postprocessor(Protocol):
    def __call__(self, data: Any, **kwargs: Any) -> None:
        ...
```

Since `(data: Any, kwarg_with_default: Any = ...) -> None` is technically compatible since the extra kwarg is optional.

Comment (Archmonger) [reply docs/src/features/hooks.md:124]: I am struggling to see the difference between your code block and what was typed above

Comment (rmorshea) [reply docs/src/features/hooks.md:124]: sorry. i corrected the code block

**Comment by Archmonger** [comment]
- Fixed the type hint
- Modified the `Postprocessor` to now require a return value.
   - The idea behind this change is, since we aren't trying to fully limit `use_query` to database calls, we have no way of ensuring that `data` is always mutable (such as a `Model` object).
- Made the default postprocessor configurable.

Since the changes are non-trivial, I will wait for one more approval.

**Comment by Archmonger** [comment]
Pushing this forward, we can do any requested changes in the version bump PR.

**Pull request merged**
